ssl/statem/statem_clnt.c

```diff
@@ -3336,7 +3336,12 @@ int tls_construct_client_certificate(SSL *s, WPACKET *pkt)
                     SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_WRITE))) {
         SSLerr(SSL_F_TLS_CONSTRUCT_CLIENT_CERTIFICATE,
                SSL_R_CANNOT_CHANGE_CIPHER);
-        goto err;
+        /*
+         * This is a fatal error, which leaves
+         * enc_write_ctx in an inconsistent state
+         * and thus ssl3_send_alert may crash.
+         */
+        return 0;
```

Comment:
> In a less perfect world, we return -1 immediately if we get out of memory,
and do not try to send an alert about it.

Unfortunately we do not live in a perfect world. Where it makes sense to do so we should send alerts - but if we fail to do so, give up.